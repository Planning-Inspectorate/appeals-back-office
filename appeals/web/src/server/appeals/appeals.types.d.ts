import { CurrentPermissionSet } from '#environment/permissions';

export interface CheckboxRadioConditionalHtmlParameter {
	html?: string;
}

export interface CheckboxItemParameter {
	value: string;
	text: string;
	conditional?: CheckboxRadioConditionalHtmlParameter;
	checked?: boolean;
}

export interface SelectItemParameter {
	value: string;
	text: string;
	checked?: boolean;
}

export interface DayMonthYearHourMinute {
	day?: number | string;
	month?: number | string; // 1-based, i.e. January === 1 (Date stores this as 0-based value, eg. Date.getMonth() called on a date in January will return 0)
	year?: number | string;
	hour?: number | string;
	minute?: number | string;
}

export type DocumentVirusCheckStatus = 'not_scanned' | 'scanned' | 'affected';
export type DocumentRowDisplayMode = 'none' | 'number' | 'list';

declare global {
	namespace Express {
		interface Request {
			currentFolder: Schema.Folder;
			currentAppeal: Appeal;
			currentComment: Representation;
			currentRepresentation: Representation;
			apiClient: import('got').Got;
			permissions: CurrentPermissionSet;
		}
	}
}
