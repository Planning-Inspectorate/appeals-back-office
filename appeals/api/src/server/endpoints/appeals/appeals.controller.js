import { getPageCount } from '#utils/database-pagination.js';
import { sortAppeals } from '#utils/appeal-sorter.js';
import appealRepository from '#repositories/appeal.repository.js';
import appealListRepository from '#repositories/appeal-lists.repository.js';
import {
	DEFAULT_PAGE_NUMBER,
	DEFAULT_PAGE_SIZE,
	ERROR_CANNOT_BE_EMPTY_STRING
} from '@pins/appeals/constants/support.js';
import { formatMyAppeal } from './appeals.formatter.js';
import { retrieveAppealListData, updateCompletedEvents } from './appeals.service.js';
import { isFeatureActive } from '#utils/feature-flags.js';
import { FEATURE_FLAG_NAMES } from '@pins/appeals/constants/common.js';

/** @typedef {import('express').Request} Request */
/** @typedef {import('express').Response} Response */
/** @typedef {import('@pins/appeals.api').Appeals.SingleAppealDetailsResponse} SingleAppealDetailsResponse */

/**
 * @param {Request} req
 * @param {Response} res
 * @returns {Promise<Response>}
 */
const getAppeals = async (req, res) => {
	const { query } = req;
	let { searchTerm, status, hasInspector, lpaCode, inspectorId, caseOfficerId } = query;
	const pageNumber = Number(query.pageNumber) || DEFAULT_PAGE_NUMBER;
	const pageSize = Number(query.pageSize) || DEFAULT_PAGE_SIZE;
	const isGreenBelt = query.isGreenBelt === 'true';
	const appealTypeId = Number(query.appealTypeId) || null;

	const {
		itemCount,
		mappedAppeals,
		mappedStatuses,
		statusesInNationalList,
		mappedLPAs,
		mappedInspectors,
		mappedCaseOfficers
	} = await retrieveAppealListData(
		pageNumber,
		pageSize,
		// @ts-ignore
		searchTerm,
		status,
		hasInspector,
		lpaCode,
		inspectorId,
		caseOfficerId,
		isGreenBelt,
		appealTypeId
	);

	return res.send({
		itemCount,
		items: mappedAppeals,
		statuses: mappedStatuses,
		statusesInNationalList,
		lpas: mappedLPAs,
		inspectors: mappedInspectors,
		caseOfficers: mappedCaseOfficers,
		page: pageNumber,
		pageCount: getPageCount(itemCount, pageSize),
		pageSize
	});
};

/**
 * @param {Request} req
 * @param {Response} res
 * @returns {Promise<Response>}
 */
const getMyAppeals = async (req, res) => {
	const { query } = req;
	const pageNumber = Number(query.pageNumber) || DEFAULT_PAGE_NUMBER;
	const pageSize = Number(query.pageSize) || DEFAULT_PAGE_SIZE;
	const status = query.status ? String(query.status) : undefined;
	const azureUserId = req.get('azureAdUserId');

	if (!azureUserId) {
		return res.status(401).send({ errors: { azureUserId: ERROR_CANNOT_BE_EMPTY_STRING } });
	}

	const [itemCount, appeals = [], statuses] = await appealListRepository.getUserAppeals(
		azureUserId,
		pageNumber,
		pageSize,
		status
	);

	const formattedAppeals = await Promise.all(
		appeals.map(async (appeal) => {
			const linkedAppeals = await appealRepository.getLinkedAppeals(appeal.reference);
			const isChildAppeal =
				isFeatureActive(FEATURE_FLAG_NAMES.LINKED_APPEALS) &&
				linkedAppeals.some((link) => link.childRef === appeal.reference);
			const isParentAppeal =
				isFeatureActive(FEATURE_FLAG_NAMES.LINKED_APPEALS) &&
				linkedAppeals.some((link) => link.parentRef === appeal.reference);
			const myAppealData = isChildAppeal ? [] : [{ appeal, isParentAppeal, isChildAppeal }];
			if (isParentAppeal) {
				await Promise.all(
					linkedAppeals.map(async (linkedAppeal) => {
						if (linkedAppeal.childId) {
							const childAppeal = await appealRepository.getAppealById(
								Number(linkedAppeal.childId)
							);
							if (childAppeal) {
								myAppealData.push({
									// @ts-ignore
									appeal: childAppeal,
									isParentAppeal: false,
									// @ts-ignore
									isChildAppeal: true
								});
							}
						}
					})
				);
			}
			return Promise.all(myAppealData.map(formatMyAppeal));
		})
	);
	const sortedAppeals = sortAppeals(formattedAppeals.flat());

	// Flatten to a unique array of strings
	// @ts-ignore
	const formattedStatuses = statuses
		// @ts-ignore
		?.map(({ appealStatus }) => appealStatus.map(({ status }) => status))
		.flat()
		.filter((status, index, statuses) => statuses.indexOf(status) === index);

	return res.send({
		itemCount,
		items: sortedAppeals,
		statuses: formattedStatuses,
		page: pageNumber,
		pageCount: getPageCount(itemCount, pageSize),
		pageSize
	});
};

/**
 * @param {Request} req
 * @param {Response} res
 * @returns {Promise<Response>}
 */
async function updateCompletedEventsController(req, res) {
	const { azureAdUserId } = req.params;

	await updateCompletedEvents(azureAdUserId);

	return res.status(204).end();
}

export { getAppeals, getMyAppeals, updateCompletedEventsController };
