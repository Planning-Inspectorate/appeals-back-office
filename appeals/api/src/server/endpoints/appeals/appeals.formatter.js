import appealRepository from '#repositories/appeal.repository.js';
import { calculateDueDate } from '#utils/calculate-due-date.js';
import { currentStatus } from '#utils/current-status.js';
import formatAddress from '#utils/format-address.js';
import { formatCostsDecision } from '#utils/format-costs-decision.js';
import {
	formatAppellantCaseDocumentationStatus,
	formatLpaQuestionnaireDocumentationStatus,
	formatLpaStatementStatus
} from '#utils/format-documentation-status.js';
import { isAwaitingLinkedAppeal } from '#utils/is-awaiting-linked-appeal.js';
import { APPEAL_REPRESENTATION_TYPE } from '@pins/appeals/constants/common.js';
import {
	DOCUMENT_STATUS_NOT_RECEIVED,
	DOCUMENT_STATUS_RECEIVED
} from '@pins/appeals/constants/support.js';
import isFPA from '@pins/appeals/utils/is-fpa.js';
import { APPEAL_CASE_STATUS } from '@planning-inspectorate/data-model';
import { countBy } from 'lodash-es';

/** @typedef {import('@pins/appeals.api').Schema.Appeal} Appeal */
/** @typedef {import('@pins/appeals.api').Schema.Folder} Folder */
/** @typedef {import('@pins/appeals.api').Schema.AppealRelationship} AppealRelationship */
/** @typedef {import('@pins/appeals.api').Schema.AppealType} AppealType */
/** @typedef {import('@pins/appeals.api').Appeals.AppealListResponse} AppealListResponse */
/** @typedef {import('@pins/appeals.api').Appeals.AppealTimetable} AppealTimetable */
/** @typedef {import('@pins/appeals.api').Appeals.SingleAppealDetailsResponse} SingleAppealDetailsResponse */
/** @typedef {import('@pins/appeals').CostsDecision} CostsDecision */
/** @typedef {import('#endpoints/appeals').DocumentationSummary} DocumentationSummary */
/** @typedef {import('#db-client').AppealStatus} AppealStatus */
/** @typedef {import('@pins/appeals.api').Schema.Representation} Representation */
/** @typedef {import('#repositories/appeal-lists.repository.js').DBAppeals} DBAppeals */
/** @typedef {DBAppeals[0]} DBAppeal */
/** @typedef {import('#repositories/appeal-lists.repository.js').DBUserAppeal} DBUserAppeal */

/**
 * @param {DBAppeal & {costsDecision?: CostsDecision}} appeal
 * @param {AppealRelationship[]} linkedAppeals
 * @returns {AppealListResponse}
 */
const formatAppeal = (appeal, linkedAppeals) => {
	return {
		appealId: appeal.id,
		appealReference: appeal.reference,
		appealSite: formatAddress(appeal.address),
		appealStatus: currentStatus(appeal),
		appealType: appeal.appealType?.type,
		procedureType: appeal.procedureType?.name,
		createdAt: appeal.caseCreatedDate,
		localPlanningDepartment: appeal.lpa?.name || '',
		dueDate: null,
		documentationSummary: formatDocumentationSummary(appeal),
		appealTimetable: formatAppealTimetable(appeal),
		isParentAppeal: linkedAppeals.filter((link) => link.parentRef === appeal.reference).length > 0,
		isChildAppeal: linkedAppeals.filter((link) => link.childRef === appeal.reference).length > 0,
		planningApplicationReference: appeal.applicationReference,
		isHearingSetup: !!appeal.hearing,
		hasHearingAddress: !!appeal.hearing?.addressId,
		awaitingLinkedAppeal: null,
		numberOfResidencesNetChange: appeal.appellantCase?.numberOfResidencesNetChange || null
	};
};

/**
 * @param {Object} options
 * @param {DBUserAppeal} options.appeal
 * @param {Boolean} [options.isParentAppeal]
 * @param {Boolean} [options.isChildAppeal]
 * @param {Boolean} [options.awaitingLinkedAppeal]
 * @returns {Promise<AppealListResponse>}
 */
const formatMyAppeal = async ({
	appeal,
	isParentAppeal = false,
	isChildAppeal = false,
	awaitingLinkedAppeal = false
}) => {
	const costsDecision = await formatCostsDecision(appeal);
	return {
		appealId: appeal.id,
		appealReference: appeal.reference,
		appealSite: formatAddress(appeal.address),
		appealStatus: currentStatus(appeal),
		appealType: appeal.appealType?.type,
		procedureType: appeal.procedureType?.name,
		createdAt: appeal.caseCreatedDate,
		localPlanningDepartment: appeal.lpa?.name || '',
		lpaQuestionnaireId: appeal.lpaQuestionnaire?.id || null,
		documentationSummary: formatDocumentationSummary(appeal),
		dueDate: await calculateDueDate(appeal, costsDecision),
		appealTimetable: formatAppealTimetable(appeal),
		isParentAppeal,
		isChildAppeal,
		planningApplicationReference: appeal.applicationReference,
		isHearingSetup: !!appeal.hearing,
		hasHearingAddress: !!appeal.hearing?.addressId,
		awaitingLinkedAppeal,
		costsDecision,
		numberOfResidencesNetChange: appeal.appellantCase?.numberOfResidencesNetChange ?? null
	};
};

/**
 *
 * @param {Object} options
 * @param {number} options.appealId
 * @param {DBUserAppeal} options.appeal
 * @param {Date} options.dueDate
 * @param {String} options.linkType
 * @param {Boolean} [options.awaitingLinkedAppeal]
 * @returns {Promise<AppealListResponse>}
 */
const formatPersonalListItem = async ({
	appealId,
	appeal,
	dueDate,
	linkType,
	awaitingLinkedAppeal = false
}) => {
	const { reference, lpaQuestionnaire, appellantCase, hearing, procedureType, appealType } = appeal;
	const appealStatus = currentStatus(appeal);
	const appealIsCompleteOrWithdrawn =
		appealStatus === APPEAL_CASE_STATUS.COMPLETE || appealStatus === APPEAL_CASE_STATUS.WITHDRAWN;

	return {
		appealId,
		appealReference: reference,
		appealSite: formatAddress(appeal.address),
		appealStatus,
		appealType: appealType?.type,
		procedureType: procedureType?.name,
		createdAt: appeal.caseCreatedDate,
		localPlanningDepartment: appeal.lpa?.name || '',
		lpaQuestionnaireId: lpaQuestionnaire?.id ?? null,
		documentationSummary: formatDocumentationSummary(appeal),
		dueDate,
		appealTimetable: formatAppealTimetable(appeal),
		isParentAppeal: linkType === 'parent',
		isChildAppeal: linkType === 'child',
		planningApplicationReference: appeal.applicationReference,
		isHearingSetup: !!hearing,
		hasHearingAddress: !!hearing?.addressId,
		awaitingLinkedAppeal,
		costsDecision: appealIsCompleteOrWithdrawn ? await formatCostsDecision(appeal) : null,
		numberOfResidencesNetChange: appellantCase?.numberOfResidencesNetChange ?? null
	};
};

/**
 * @param {DBAppeal | DBUserAppeal} appeal
 * @returns {DocumentationSummary}
 * */
const formatDocumentationSummary = (appeal) => {
	const ipComments =
		appeal.representations?.filter(
			(rep) => rep.representationType === APPEAL_REPRESENTATION_TYPE.COMMENT
		) ?? [];

	const lpaStatement = appeal.representations?.find(
		(rep) => rep.representationType === APPEAL_REPRESENTATION_TYPE.LPA_STATEMENT
	);
	const lpaFinalComments =
		appeal.representations?.filter(
			(rep) => rep.representationType === APPEAL_REPRESENTATION_TYPE.LPA_FINAL_COMMENT
		) ?? [];
	const appellantFinalComments =
		appeal.representations?.filter(
			(rep) => rep.representationType === APPEAL_REPRESENTATION_TYPE.APPELLANT_FINAL_COMMENT
		) ?? [];

	return {
		appellantCase: {
			status: formatAppellantCaseDocumentationStatus(appeal),
			dueDate: appeal.caseExtensionDate && appeal.caseExtensionDate?.toISOString(),
			receivedAt: appeal.caseCreatedDate?.toISOString()
		},
		lpaQuestionnaire: {
			status: formatLpaQuestionnaireDocumentationStatus(appeal),
			dueDate:
				appeal.appealTimetable?.lpaQuestionnaireDueDate &&
				appeal.appealTimetable?.lpaQuestionnaireDueDate.toISOString(),
			receivedAt:
				appeal.lpaQuestionnaire?.lpaqCreatedDate &&
				appeal.lpaQuestionnaire?.lpaqCreatedDate.toISOString()
		},
		ipComments: {
			status: ipComments.length > 0 ? DOCUMENT_STATUS_RECEIVED : DOCUMENT_STATUS_NOT_RECEIVED,
			counts: countBy(ipComments, 'status'),
			isRedacted: ipComments.some((comment) => Boolean(comment.redactedRepresentation))
		},
		lpaStatement: {
			status: formatLpaStatementStatus(lpaStatement ?? null),
			representationStatus: lpaStatement?.status ?? null,
			receivedAt: lpaStatement?.dateCreated,
			isRedacted: Boolean(lpaStatement?.redactedRepresentation)
		},
		lpaFinalComments: {
			status: lpaFinalComments.length > 0 ? DOCUMENT_STATUS_RECEIVED : DOCUMENT_STATUS_NOT_RECEIVED,
			receivedAt: lpaFinalComments[0]?.dateCreated
				? lpaFinalComments[0].dateCreated.toISOString()
				: null,
			representationStatus: lpaFinalComments[0]?.status ?? null,
			isRedacted: lpaFinalComments.some((comment) => Boolean(comment.redactedRepresentation))
		},
		appellantFinalComments: {
			status:
				appellantFinalComments.length > 0 ? DOCUMENT_STATUS_RECEIVED : DOCUMENT_STATUS_NOT_RECEIVED,
			receivedAt: appellantFinalComments[0]?.dateCreated
				? appellantFinalComments[0].dateCreated.toISOString()
				: null,
			representationStatus: appellantFinalComments[0]?.status ?? null,
			isRedacted: appellantFinalComments.some((comment) => Boolean(comment.redactedRepresentation))
		}
	};
};

/**
 * @param {DBAppeal | DBUserAppeal} appeal
 * @returns {AppealTimetable | undefined}
 * */
function formatAppealTimetable(appeal) {
	if (!appeal.appealTimetable) {
		return undefined;
	}

	return {
		appealTimetableId: appeal.appealTimetable.id,
		lpaQuestionnaireDueDate: appeal.appealTimetable.lpaQuestionnaireDueDate?.toISOString() || null,
		caseResubmissionDueDate: appeal.appealTimetable.caseResubmissionDueDate?.toISOString() || null,
		...(isFPA(appeal.appealType?.key || '') && {
			ipCommentsDueDate: appeal.appealTimetable.ipCommentsDueDate?.toISOString() || null,
			appellantStatementDueDate:
				appeal.appealTimetable.appellantStatementDueDate?.toISOString() || null,
			lpaStatementDueDate: appeal.appealTimetable.lpaStatementDueDate?.toISOString() || null,
			finalCommentsDueDate: appeal.appealTimetable.finalCommentsDueDate?.toISOString() || null,
			s106ObligationDueDate: appeal.appealTimetable.s106ObligationDueDate?.toISOString() || null,
			issueDeterminationDate: appeal.appealTimetable.issueDeterminationDate?.toISOString() || null
		})
	};
}

/**
 * @param {AppealRelationship[]} otherAppeals
 * @param {string} currentAppealRef
 * @returns {number[]}
 */
const getIdsOfReferencedAppeals = (otherAppeals, currentAppealRef) => {
	/**
	 * @type {number[]}
	 */
	const relevantIds = [];
	otherAppeals.map((relation) => {
		if (
			relation.childRef === currentAppealRef &&
			relation.parentId &&
			relevantIds.indexOf(relation.parentId) === -1
		) {
			relevantIds.push(relation.parentId);
		}
		if (
			relation.parentRef === currentAppealRef &&
			relation.childId &&
			relevantIds.indexOf(relation.childId) === -1
		) {
			relevantIds.push(relation.childId);
		}
	});

	return relevantIds;
};

/**
 *
 * @param {DBUserAppeal} appeal
 * @param {AppealRelationship[]} linkedAppeals
 * @param {boolean} isParentAppeal
 * @param {boolean} isChildAppeal
 * @returns {Promise<*[]|[{appeal, isParentAppeal, isChildAppeal, awaitingLinkedAppeal}]>}
 */
const formatLinkedAppealData = async function (
	appeal,
	linkedAppeals,
	isParentAppeal,
	isChildAppeal
) {
	// Do not add child appeals here as they will be grouped with their parent appeals below
	const myAppealData = isChildAppeal
		? []
		: [{ appeal, isParentAppeal, isChildAppeal, awaitingLinkedAppeal: false }];
	if (isParentAppeal) {
		const childAppeals = isChildAppeal
			? []
			: await Promise.all(
					linkedAppeals.map(async (linkedAppeal) => {
						if (linkedAppeal.childId) {
							const childAppeal = await appealRepository.getAppealById(
								Number(linkedAppeal.childId)
							);
							return {
								// @ts-ignore
								appeal: childAppeal,
								isParentAppeal: false,
								// @ts-ignore
								isChildAppeal: true
							};
						}
					})
			  );
		if (childAppeals?.length) {
			// @ts-ignore
			myAppealData.push(...childAppeals);
		}
	}
	return myAppealData
		.filter((appealData) => appealData)
		.map((appealData) => ({
			...appealData,
			awaitingLinkedAppeal: isAwaitingLinkedAppeal(appealData.appeal, myAppealData)
		}));
};

export {
	formatAppeal,
	formatDocumentationSummary,
	formatLinkedAppealData,
	formatMyAppeal,
	formatPersonalListItem,
	getIdsOfReferencedAppeals
};
