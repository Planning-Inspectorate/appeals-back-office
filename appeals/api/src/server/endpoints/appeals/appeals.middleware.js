import { ERROR_MUST_BE_BOOLEAN, ERROR_MUST_BE_VALID_APPEAL_STATE } from '#endpoints/constants.js';
import { APPEAL_CASE_STATUS } from 'pins-data-model';

/** @typedef {import('express').Request} Request */
/** @typedef {import('express').Response} Response */
/** @typedef {import('express').NextFunction} NextFunction */

/**
 * @type {import("express").RequestHandler}
 * @returns {Promise<object|void>}
 */
export const validateHasInspector = async (req, res, next) => {
	const hasInspector = req.query.hasInspector;

	if (hasInspector !== undefined && hasInspector !== 'true' && hasInspector !== 'false') {
		return res.status(400).send({ errors: { hasInspector: ERROR_MUST_BE_BOOLEAN } });
	}

	next();
};

/**
 * @type {import("express").RequestHandler}
 * @returns {Promise<object|void>}
 */
export const validateAppealStatus = async (req, res, next) => {
	const status = req.query.status;
	const validStatuses = [
		APPEAL_CASE_STATUS.ASSIGN_CASE_OFFICER,
		APPEAL_CASE_STATUS.VALIDATION,
		APPEAL_CASE_STATUS.READY_TO_START,
		APPEAL_CASE_STATUS.LPA_QUESTIONNAIRE,
		APPEAL_CASE_STATUS.ISSUE_DETERMINATION,
		APPEAL_CASE_STATUS.COMPLETE,
		APPEAL_CASE_STATUS.INVALID,
		APPEAL_CASE_STATUS.WITHDRAWN,
		APPEAL_CASE_STATUS.CLOSED,
		APPEAL_CASE_STATUS.AWAITING_TRANSFER,
		APPEAL_CASE_STATUS.TRANSFERRED,
		APPEAL_CASE_STATUS.EVENT,
		APPEAL_CASE_STATUS.AWAITING_EVENT,
		//S78
		APPEAL_CASE_STATUS.FINAL_COMMENTS,
		APPEAL_CASE_STATUS.STATEMENTS
	];

	const isValidStatus = typeof status === 'string' ? validStatuses.includes(status) : !status;

	if (!isValidStatus) {
		return res.status(400).send({ errors: { status: ERROR_MUST_BE_VALID_APPEAL_STATE } });
	}
	next();
};
