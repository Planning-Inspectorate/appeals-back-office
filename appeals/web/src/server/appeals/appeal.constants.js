/**
 * @typedef {'appealDetails'|'appellantCase'|'lpaQuestionnaire'|'manageDocuments'|'manageFolder'|'manageRelatedAppeals'|'manageNeighbouringSites'|'ipComments'|'viewIpComment'} ServicePageName
 */

export const paginationDefaultSettings = {
	pageSize: 30,
	firstPageNumber: 1
};

export const textInputCharacterLimits = {
	defaultAddressInputLength: 250,
	defaultInputLength: 300,
	defaultTextareaLength: 1000
};

export const kilobyte = 1000;
export const megabyte = 1000000;
export const gigabyte = 1000000000;
