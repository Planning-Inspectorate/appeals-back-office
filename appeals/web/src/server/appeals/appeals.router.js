import { Router as createRouter } from 'express';
import nationalListRouter from '../appeals/national-list/national-list.router.js';
import personalListRouter from '../appeals/personal-list/personal-list.router.js';
import appealDetailsRouter from './appeal-details/appeal-details.router.js';
import { assertUserHasPermission } from '#app/auth/auth.guards.js';
import { permissionNames } from '#environment/permissions.js';

const router = createRouter();

router.use(
	'/appeals-list',
	assertUserHasPermission(permissionNames.viewCaseList),
	nationalListRouter
);

router.use('/personal-list', personalListRouter);

router.use('/appeal-details', appealDetailsRouter);

export default router;
