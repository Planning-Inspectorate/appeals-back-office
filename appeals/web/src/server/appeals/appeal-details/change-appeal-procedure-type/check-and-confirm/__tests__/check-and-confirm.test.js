import { parseHtml } from '@pins/platform';
import nock from 'nock';
import supertest from 'supertest';
import { createTestEnvironment } from '#testing/index.js';
import { appealData } from '#testing/appeals/appeals.js';

const { app, teardown } = createTestEnvironment();
const request = supertest(app);
const appealDataWithoutStartDate = {
	...appealData,
	startedAt: null,
	appealTimetable: {
		appealTimetableId: 1053,
		lpaQuestionnaireDueDate: '2023-10-11T01:00:00.000Z',
		finalCommentsDueDate: '2023-10-12T01:00:00.000Z',
		ipCommentsDueDate: '2023-10-13T01:00:00.000Z',
		lpaStatementDueDate: '2023-10-14T01:00:00.000Z'
	}
};

describe('GET /change-appeal-procedure-type/check-and-confirm', () => {
	afterEach(teardown);

	it('should render the check details page with the expected content for written procedure type if an appeal procedure is found in the session', async () => {
		nock('http://test/')
			.get('/appeals/1')
			.reply(200, {
				...appealDataWithoutStartDate,
				appealStatus: 'lpa_questionnaire',
				appealType: 'Planning appeal',
				documentationSummary: {
					lpaQuestionnaire: {
						status: 'not received'
					}
				}
			});

		const response = await request.get(
			'/appeals-service/appeal-details/1/change-appeal-procedure-type/check-and-confirm'
		);

		expect(response.statusCode).toBe(200);

		const html = parseHtml(response.text).innerHTML;

		expect(html).toMatchSnapshot();

		const unprettifiedHtml = parseHtml(response.text, { skipPrettyPrint: true }).innerHTML;

		expect(unprettifiedHtml).toContain('Check details and update appeal procedure</h1>');
		expect(unprettifiedHtml).toContain('Appeal procedure</dt>');
		expect(unprettifiedHtml).toContain('Written representations</dd>');
		expect(unprettifiedHtml).toContain(
			'<a class="govuk-link" href="/appeals-service/appeal-details/1/change-appeal-procedure-type/change-selected-procedure-type" data-cy="change-appeal-procedure">Change<span class="govuk-visually-hidden"> Appeal procedure</span></a>'
		);
		expect(unprettifiedHtml).toContain('<h3 class="govuk-heading-m">Timetable due dates</h3>');

		expect(unprettifiedHtml).toContain('Interested party comments due</dt>');
		expect(unprettifiedHtml).toContain('13 October 2023</dd>');
		expect(unprettifiedHtml).toContain('<a class="govuk-link" href="change-timetable">Change ');

		expect(unprettifiedHtml).toContain('Statements due</dt>');
		expect(unprettifiedHtml).toContain('14 October 2023</dd>');
		expect(unprettifiedHtml).toContain('<a class="govuk-link" href="change-timetable">Change ');

		expect(unprettifiedHtml).toContain('LPA questionnaire due</dt>');
		expect(unprettifiedHtml).toContain('11 October 2023</dd>');
		expect(unprettifiedHtml).toContain('<a class="govuk-link" href="change-timetable">Change ');

		expect(unprettifiedHtml).toContain('Final comments due</dt>');
		expect(unprettifiedHtml).toContain('12 October 2023</dd>');
		expect(unprettifiedHtml).toContain('<a class="govuk-link" href="change-timetable">Change ');

		expect(unprettifiedHtml).toContain(
			`Weâ€™ll send an email to the appellant and LPA to tell them that:</p>`
		);
		expect(unprettifiedHtml).toContain(`we've changed the procedure`);
		expect(unprettifiedHtml).toContain(`we've cancelled the hearing`);
		expect(unprettifiedHtml).toContain('Update appeal procedure</button>');
	});
});
