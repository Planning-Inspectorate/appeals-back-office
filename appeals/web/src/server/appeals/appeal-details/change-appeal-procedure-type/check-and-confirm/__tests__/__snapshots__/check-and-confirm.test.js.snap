// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /change-appeal-procedure-type/check-and-confirm should render the check details page with the expected content for written procedure type if an appeal procedure is found in the session 1`] = `
"<main class="govuk-main-wrapper" id="main-content">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds"><span class="govuk-caption-l">Appeal APP/Q9999/D/21/351062</span>
            <h1 class="govuk-heading-l">Check details and update appeal procedure</h1>
        </div>
    </div>
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <dl class="govuk-summary-list">
                <div class="govuk-summary-list__row"><dt class="govuk-summary-list__key"> Appeal procedure</dt>
                    <dd class="govuk-summary-list__value">Written representations</dd>
                    <dd class="govuk-summary-list__actions"><a class="govuk-link" href="/appeals-service/appeal-details/1/change-appeal-procedure-type/change-selected-procedure-type"
                        data-cy="change-appeal-procedure">Change<span class="govuk-visually-hidden"> Appeal procedure</span></a>
                    </dd>
                </div>
            </dl>
            <h3 class="govuk-heading-m">Timetable due dates</h3>
            <dl class="govuk-summary-list">
                <div class="govuk-summary-list__row"><dt class="govuk-summary-list__key"> LPA questionnaire due</dt>
                    <dd class="govuk-summary-list__value">11 October 2023</dd>
                    <dd class="govuk-summary-list__actions"><a class="govuk-link" href="change-timetable">Change <span class="govuk-visually-hidden">LPA questionnaire due date</span></a>
                    </dd>
                </div>
                <div class="govuk-summary-list__row"><dt class="govuk-summary-list__key"> Statements due</dt>
                    <dd class="govuk-summary-list__value">14 October 2023</dd>
                    <dd class="govuk-summary-list__actions"><a class="govuk-link" href="change-timetable">Change <span class="govuk-visually-hidden">Statements due date</span></a>
                    </dd>
                </div>
                <div class="govuk-summary-list__row"><dt class="govuk-summary-list__key"> Interested party comments due</dt>
                    <dd                     class="govuk-summary-list__value">13 October 2023</dd>
                        <dd class="govuk-summary-list__actions"><a class="govuk-link" href="change-timetable">Change <span class="govuk-visually-hidden">Interested party comments due date</span></a>
                        </dd>
                </div>
                <div class="govuk-summary-list__row"><dt class="govuk-summary-list__key"> Final comments due</dt>
                    <dd class="govuk-summary-list__value">12 October 2023</dd>
                    <dd class="govuk-summary-list__actions"><a class="govuk-link" href="change-timetable">Change <span class="govuk-visually-hidden">Final comments due date</span></a>
                    </dd>
                </div>
            </dl>
            <p class="govuk-body">Weâ€™ll send an email to the appellant and LPA to tell them that:</p>
            <li             class="govuk-body">we've changed the procedure</li>
                <li class="govuk-body">we've cancelled the hearing</li>
                <div class="govuk-grid-row">
                    <div class="govuk-grid-column-full">
                        <form method="post" novalidate="novalidate">
                            <button type="submit" class="govuk-button" data-module="govuk-button">Update appeal procedure</button>
                        </form>
                    </div>
                </div>
        </div>
    </div>
</main>"
`;
